examples/syscall.c

```diff
@@ -83,37 +113,91 @@ asmlinkage int our_sys_open(const char *filename, int flags, int mode)
 
 static unsigned long **aquire_sys_call_table(void)
 {
+#ifdef HAVE_KSYS_CLOSE
     unsigned long int offset = PAGE_OFFSET;
     unsigned long **sct;
 
     while (offset < ULLONG_MAX) {
         sct = (unsigned long **) offset;
 
-        if (sct[__NR_close] == (unsigned long *) close_fd)
+        if (sct[__NR_close] == (unsigned long *) ksys_close)
             return sct;
 
         offset += sizeof(void *);
     }
 
     return NULL;
+#endif
+
+#ifdef HAVE_PARAM
+    const char sct_name[15] = "sys_call_table";
+    char symbol[40] = {0};
+
+    if (sym == 0) {
+        pr_alert(
+            "For Linux v5.7+, Kprobes is the preferable way to get "
+            "symbol.\n");
+        pr_info(
+            "If Kprobes is absent, you have to specify the address of "
+            "sys_call_table from\n");
+        pr_info("/boot/System.map or /proc/kallsyms into sym parameter.\n");
+        return NULL;
+    }
+    sprint_symbol(symbol, sym);
+    if (!strncmp(sct_name, symbol, sizeof(sct_name) - 1))
+        return (unsigned long **) sym;
+
+    return NULL;
+#endif
+
+#ifdef HAVE_KPROBES
+    unsigned long (*kallsyms_lookup_name)(const char *name);
+    if (register_kprobe(&kp) < 0)
+        return NULL;
+    kallsyms_lookup_name = (unsigned long (*)(const char *name)) kp.addr;
+    unregister_kprobe(&kp);
+#endif
+
+    return (unsigned long **) kallsyms_lookup_name("sys_call_table");
+}
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 3, 0)
+static inline void __write_cr0(unsigned long cr0)
```

Comment:
You can safely rename the function to `write_cr0` since it is a static inline one.

Reply:
I get the error message when using `write_cr0`:
```cpp
164 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 3, 0)                                                                                                                                                                                                         
165 static inline void write_cr0(unsigned long cr0)
166 {
167     asm volatile("mov %0,%%cr0" : "+r"(cr0) : : "memory");
168 }
169 #endif
```
```shell
/home/slda/lkmpg/examples/syscall.c:165:20: error: redefinition of ‘write_cr0’
  165 | static inline void write_cr0(unsigned long cr0)
      |                    ^~~~~~~~~
In file included from ./arch/x86/include/asm/msr.h:255,
                 from ./arch/x86/include/asm/processor.h:22,
                 from ./arch/x86/include/asm/timex.h:5,
                 from ./include/linux/timex.h:65,
                 from ./include/linux/time32.h:13,
                 from ./include/linux/time.h:60,
                 from ./include/linux/stat.h:19,
                 from ./include/linux/module.h:13,
                 from /home/slda/lkmpg/examples/syscall.c:15:
./arch/x86/include/asm/paravirt.h:118:20: note: previous definition of ‘write_cr0’ was here
  118 | static inline void write_cr0(unsigned long x)
```

Reply:
Can we re-use `write_cr0`?

Reply:
Cannot, this is one of the reasons why we have this commit.
See [commit](https://github.com/torvalds/linux/commit/8dbec27a242cd3e2816eeb98d3237b9f57cf6232).